### PR TITLE
Adding percentage of spread & Resolve missing hours issue

### PR DESCRIPTION
**Pull request by Pluimvee**
I've added a percentage of the spread instead of a percentage of the max price. Using the percentage its easier to determine a low-price period to which I can turn on-/off some devices.

The calculation is based on the same data set as the max/min/average/percentage_of_max

### CODE REVIEW

**Comment by Roeland54** [comment]
Your pr reverts the changes of #188 that is probably not intentional.

**Comment by Pluimvee** [comment]
Indeed, my mistake.

**Comment by Pluimvee** [comment]
Reapplied changes to the latest version

File: custom_components/entsoe/sensor.py [hunk 1]
```diff
@@ -105,6 +105,15 @@ def sensor_descriptions(
             state_class=SensorStateClass.MEASUREMENT,
             value_fn=lambda coordinator: coordinator.get_percentage_of_max(),
         ),
+        EntsoeEntityDescription(
+            key="percentage",
+            name="Current percentage",
```
Comment (Roeland54) [inline custom_components/entsoe/sensor.py:110]: I would like a more descriptive name.

**Review by Roeland54 — CHANGES_REQUESTED** [review]

File: custom_components/entsoe/api_client.py [hunk 2]
```diff
@@ -96,7 +96,21 @@ def query_day_ahead_prices(
...
                             series[date + timedelta(hours=hour)] = float(price)
 
+                # Start filling from the first hour in the series until the requested end
+                current_time = min(series.keys(), default=None)  # Start from the first time in response
+                if current_time is None:
+                    return series  # No valid data received from the server
+
+                last_price = series[current_time]
+
+                while current_time < end:
```
[the remark this answers is not in the document]
Comment (Pluimvee) [reply custom_components/entsoe/api_client.py:106]: You are correct

File: custom_components/entsoe/sensor.py (hunk 1 above)
Comment (Pluimvee) [reply custom_components/entsoe/sensor.py:110]: Any suggestions?

File: custom_components/entsoe/coordinator.py
Comment (Pluimvee) [inline custom_components/entsoe/coordinator.py:1]: Just added some comments to explain what happens here. I plan to make some changes to the sliding code to limit to 24 hrs. Also thinking of adding another sliding loop over 12 hrs to detect the low pricing periods around 3:00 and 13:00 o'clock

Comment (Roeland54) [reply custom_components/entsoe/coordinator.py:1]: I think the current assortment of calculation methods is already confusing enough and it is my opinion that the focus of this integration should be on fetching the correct price in a stable way. Which is already bigger challenge than it should be. The integration provides an [action](https://github.com/JaccoR/hass-entso-e/issues/147) that makes it possible for other home assistant components to easily fetch the price. More calculations based on price should be a separate thing like for example: [cheapest-energy-hours macro](https://github.com/TheFes/cheapest-energy-hours).

File: custom_components/entsoe/coordinator.py [hunk 3]
```diff
@@ -190,17 +190,21 @@ def today_data_available(self):
...
             return {
                 hour: price
                 for hour, price in data.items()
                 if hour >= self.today and hour < self.today + timedelta(days=1)
             }
+        # sliding = calculations made on all data from the current hour and beyond (future data only)
         elif self.calculation_mode == CALCULATION_MODE["sliding"]:
             now = dt.now().replace(minute=0, second=0, microsecond=0)
             return {hour: price for hour, price in data.items() if hour >= now}
+        # rotation = calculations made on all data of today and tomorrow (48 hrs) if we have 72hrs of data
```
Comment (Roeland54) [inline custom_components/entsoe/coordinator.py:204]: Did you mean publish here?

File: custom_components/entsoe/coordinator.py [hunk 4]
```diff
@@ -190,17 +190,21 @@ def today_data_available(self):
...
         elif self.calculation_mode == CALCULATION_MODE["publish"] and len(data) > 48:
             return {hour: price for hour, price in data.items() if hour >= self.today}
+        # rotation = calculations made on all data of yesterday and today (48 hrs)
```
Comment (Roeland54) [inline custom_components/entsoe/coordinator.py:207]: Did you mean publish here?

**Review by Roeland54 — CHANGES_REQUESTED** [review]

File: custom_components/entsoe/coordinator.py
Comment (Pluimvee) [reply custom_components/entsoe/coordinator.py:1]: In that case we should remove all calculations, the min, max, avg, etc

Comment (Pluimvee) [reply custom_components/entsoe/coordinator.py:1]: I will create another issue (improvement) for this as its not related to this issue

File: custom_components/entsoe/coordinator.py (hunk 3 above)
Comment (Pluimvee) [reply custom_components/entsoe/coordinator.py:204]: Yes

File: custom_components/entsoe/coordinator.py (hunk 4 above)
Comment (Pluimvee) [reply custom_components/entsoe/coordinator.py:207]: Yes

File: custom_components/entsoe/api_client.py [hunk 5]
```diff
@@ -96,7 +96,26 @@ def query_day_ahead_prices(
...
+                last_response_time = max(series.keys(), default=None)
+
+                if first_response_time is None or last_response_time is None:
+                    return series  # No valid data received from the server
+
+                # Now only fill in missing hours between the first and last response times
+                current_time = first_response_time
+                last_price = series[current_time]
+
+                while current_time <= last_response_time:
```
[the remark this answers is not in the document]
Comment (Roeland54) [reply custom_components/entsoe/api_client.py:110]: The xml response contains a start and an end timestamp for each period send. So this end time could be used to determine the correct end. 
image

[the remark this answers is not in the document]
Comment (Pluimvee) [reply custom_components/entsoe/api_client.py]: Changed the code accordingly.

Still some validation may be added here to verify if start time in message corresponds with first entry in series and is in line with the requested start time of the method

File: custom_components/entsoe/coordinator.py (hunk 3 above)
Comment (Pluimvee) [reply custom_components/entsoe/coordinator.py:204]: I cant update these comments as I'm in the middle of adding/changing the calculation modes and running tests. Please ignore these commentary hickups

File: custom_components/entsoe/coordinator.py (hunk 4 above)
Comment (Pluimvee) [reply custom_components/entsoe/coordinator.py:207]: I cant update these comments as I'm in the middle of adding/changing the calculation modes and running tests. Please ignore these commentary hiccups

Comment (Roeland54) [reply custom_components/entsoe/coordinator.py:207]: I won't merge a PR with know errors in comments. You should separate your work into branches so you can switch between them.

File: custom_components/entsoe/coordinator.py (hunk 3 above)
Comment (Roeland54) [reply custom_components/entsoe/coordinator.py:204]: I won't merge a PR with know errors in comments. You should separate your work into branches so you can switch between them.

File: custom_components/entsoe/api_client.py [hunk 6]
```diff
@@ -82,10 +82,16 @@ def query_day_ahead_prices(
...
+                        response_start = period.find(".//timeInterval/start").text
+                        start_time = (
+                            datetime.strptime(response_start, "%Y-%m-%dT%H:%MZ")
+                            .replace(tzinfo=pytz.UTC)
+                            .astimezone()
+                        )
 
-                        date = (
-                            datetime.strptime(start_time, "%Y-%m-%dT%H:%MZ")
+                        response_end = period.find(".//timeInterval/start").text
```
Comment (Roeland54) [inline custom_components/entsoe/api_client.py:92]: This looks like a copy paste error. You are fetching the 'start' tag for the reponse_end time.

**Review by Roeland54 — CHANGES_REQUESTED** [review]

File: custom_components/entsoe/api_client.py (hunk 6 above)
Comment (Pluimvee) [reply custom_components/entsoe/api_client.py]: Resolved,
I also found that the endtime should be excluded [from - end> as this value is given in a follow up period

image

**Review by Roeland54 — APPROVED** [review]

**Pull request merged**